### DISCUSSION TITLE
Negatable Signals

### DISCUSSION THREAD

**Discussion by penenkel** [datalust/seq-tickets#1104] [Ideas]
I often use signals to group events of a certain type/category/use-case. Depending on what I'm working on, I then want to see either *only* those events belonging to the signal or all events *excluding* those.

As far as I can tell there is currently no way to apply a signal negated, so that it excludes all events that it matches. 

I would appreciate if such a functionality could be added.

**Comment by nblumhardt** [comment]
Thanks for the suggestion. We've had this in the backs of our minds for some time 👍 

The main implementation challenge is to find an obvious/accessible UI for it. Some options like double-clicking to exclude suffer from discoverability problems; a "-" somewhere on the signal itself would be confusing without a "+", and within signal groups (where signals are generally mutually exclusive, and selections are unioned,) deselecting a signal is equivalent to excluding it, as long as another signal within the group is already selected, so there's some potential quirkiness to consider.

Ideas most welcome!

**Comment by nblumhardt** [comment]
Implementing this may have the side-effect of covering #860 as well, since retention policies are based on signal expressions. 🤔

**Comment by penenkel** [comment]
Has there been any development with regard to this topic? I still miss the ability to negate signals. And I am regularly annoyed about the, in my opinion misuse, of the selection mechanism for signal activation. Especially because entries are also highlighted when they are being edited. 

My opinions
- make signal activation explicit, i.e. not via selection 
- make signals negateable 
- make signal groups negateable

For negation of groups follow the usual rules of boolean algebra. That is, as grouped signals are combined via or/union by default, negated groups should be combined via and/intersection. 

Effectively each signal can have three states: inactive, active and negated.
I have to admit that I do not know of any suitable three-state control. Therefore I would probably start with a switch or checkbox control for activation and a toggle button for negation. The later perhaps utilizing the U+00AC negation symbol. Although I am having trouble finding a symbol for not-negated, there isn't even a word for it, perhaps a crossed out negation symbol or just a blank.

An other idea would utilizing set theory symbols, though those are probably even less known than the boolean algebra ones.

**Reply by nblumhardt** [reply]
Thanks for the suggestions! There hasn't been any change on this one, yet, but we're looping back to another round of planning shortly, and will take a fresh look at this.

Best regards,
Nick

**Comment by AndrewSav** [comment]
@nblumhardt Just to check in to indicate that there is still interest. I was just queried today by one of out development teams on the topic. How was that round of planning? Did this feature make it? Reference to prior discussions: https://github.com/datalust/seq-tickets/issues/860

**Reply by nblumhardt** [reply]
Thanks for the nudge, @AndrewSav. Definitely still interest but no change to report yet.
